File: MasteringSwift/MasteringSwift/chapters/chapter13/advanced/AdvancedConcurrency.swift
//
//  AdvancedConcurrency.swift
//  MasteringSwift
//
//  Capítulo 13 — Concurrencia y asincronía.
//  Nivel AVANZADO: UserService — fetchUser, fetchUserPosts, fetchUserComments; combinar resultados.
//
//  Llamar desde la app: Task { await runAdvancedConcurrencyExamples() }
//

import Foundation

fileprivate enum UserService {
    static func fetchUser() async -> String {
        try? await Task.sleep(nanoseconds: 80_000_000)
        return "User(id=1, name=Alice)"
    }

    static func fetchUserPosts(userId: Int) async -> [String] {
        try? await Task.sleep(nanoseconds: 100_000_000)
        return ["Post 1", "Post 2"]
    }

    static func fetchUserComments(userId: Int) async -> [String] {
        try? await Task.sleep(nanoseconds: 120_000_000)
        return ["Comment A", "Comment B"]
    }
}

func runAdvancedConcurrencyExamples() async {
    print("========================================")
    print("       ASYNC USER WORKFLOW")
    print("========================================")

    let user = await UserService.fetchUser()
    print("User: \(user)")

    // --- Ejecutar posts y comments en paralelo ---
    async let posts = UserService.fetchUserPosts(userId: 1)
    async let comments = UserService.fetchUserComments(userId: 1)
    let (postsResult, commentsResult) = await (posts, comments)
    print("Posts: \(postsResult)")
    print("Comments: \(commentsResult)")
    print("----------------------------------------")
    print("Done.")
    print("========================================")
}

File: MasteringDart/lib/chapters/chapter13/advanced/advanced_concurrency.dart
// ignore_for_file: avoid_print
//
//  advanced_concurrency.dart
//  MasteringDart
//
//  Capítulo 13 — Concurrencia y asincronía.
//  Nivel AVANZADO: UserService — fetchUser, fetchUserPosts, fetchUserComments; combinar resultados.
//
//  Ejecutar con: dart run lib/chapters/chapter13/advanced/advanced_concurrency.dart

class UserService {
  static Future<String> fetchUser() async {
    await Future.delayed(const Duration(milliseconds: 80));
    return 'User(id=1, name=Alice)';
  }

  static Future<List<String>> fetchUserPosts(int userId) async {
    await Future.delayed(const Duration(milliseconds: 100));
    return ['Post 1', 'Post 2'];
  }

  static Future<List<String>> fetchUserComments(int userId) async {
    await Future.delayed(const Duration(milliseconds: 120));
    return ['Comment A', 'Comment B'];
  }
}

Future<void> runAdvancedConcurrencyExamples() async {
  print('========================================');
  print('       ASYNC USER WORKFLOW');
  print('========================================');

  final user = await UserService.fetchUser();
  print('User: $user');

  // --- Ejecutar posts y comments en paralelo ---
  final postsFuture = UserService.fetchUserPosts(1);
  final commentsFuture = UserService.fetchUserComments(1);
  final posts = await postsFuture;
  final comments = await commentsFuture;
  print('Posts: $posts');
  print('Comments: $comments');
  print('----------------------------------------');
  print('Done.');
  print('========================================');
}

void main() async {
  await runAdvancedConcurrencyExamples();
}
